controllers/machine_controller.go

```diff
@@ -604,6 +621,12 @@ func (r *MachineReconciler) reconcileDeleteExternal(ctx context.Context, m *clus
 				"failed to delete %v %q for Machine %q in namespace %q",
 				obj.GroupVersionKind(), obj.GetName(), m.Name, m.Namespace)
 		}
+		// If delete is successful, then return empty object.
+		obj, err = external.Get(ctx, r.Client, ref, m.Namespace)
+		if err != nil && !apierrors.IsNotFound(errors.Cause(err)) {
+			return nil, errors.Wrapf(err, "failed to get %s %q for Machine %q in namespace %q",
+				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+		}
```

Comment:
might be I'm missing something...

> // If delete is successful, then return empty object.

At this stage I assume delete is successful, so why are we reading again the object after delete?

Reply:
Because it actually deletes external object here but non-nil object causes requeue: 
https://github.com/kubernetes-sigs/cluster-api/blob/af6630920560ca0e12179897b96d6ea8bd830b63/controllers/machine_controller.go#L573-L576

We already know the object is gone so this is to set the obj to nil.

Reply:
if the problem is to set the object to nil, should we set obj == nil without calling Get?